solr/core/src/java/org/apache/solr/cluster/maintenance/InactiveShardRemover.java

```diff
@@ -0,0 +1,219 @@
+  }
+
+  @Override
+  public void configure(final InactiveShardRemoverConfig cfg) {
+    Objects.requireNonNull(cfg, "config must be specified");
+    cfg.validate();
+    this.scheduleIntervalSeconds = cfg.scheduleIntervalSeconds;
+    this.maxDeletesPerCycle = cfg.maxDeletesPerCycle;
+    this.ttlSeconds = cfg.ttlSeconds;
+  }
+
+  @Override
+  public String getName() {
+    return PLUGIN_NAME;
+  }
+
+  @Override
+  public State getState() {
+    return state;
+  }
+
+  @Override
+  public void start() throws Exception {
+    state = State.STARTING;
+    executor = Executors.newScheduledThreadPool(1, new SolrNamedThreadFactory(PLUGIN_NAME));
+    executor.scheduleAtFixedRate(
+        this::deleteInactiveSlices,
+        scheduleIntervalSeconds,
+        scheduleIntervalSeconds,
+        TimeUnit.SECONDS);
+    state = State.RUNNING;
+  }
+
+  @Override
+  public void stop() {
+    if (state == State.RUNNING) {
+      state = State.STOPPING;
+      ExecutorUtil.shutdownNowAndAwaitTermination(executor);
+    }
+    state = State.STOPPED;
+  }
+
+  @VisibleForTesting
+  void deleteInactiveSlices() {
+    final ClusterState clusterState = coreContainer.getZkController().getClusterState();
+    Collection<Slice> inactiveSlices =
+        clusterState.getCollectionsMap().values().stream()
+            .flatMap(v -> collectInactiveSlices(v).stream())
+            .collect(Collectors.toSet());
+
+    if (log.isInfoEnabled()) {
+      log.info(
+          "Found {} inactive Shards to delete, {} will be deleted",
+          inactiveSlices.size(),
+          Math.min(inactiveSlices.size(), maxDeletesPerCycle));
+    }
+
+    inactiveSlices.stream().limit(maxDeletesPerCycle).forEach(this::deleteShard);
+  }
+
+  private Collection<Slice> collectInactiveSlices(final DocCollection docCollection) {
+    final Collection<Slice> slices = new HashSet<>(docCollection.getSlices());
+    slices.removeAll(docCollection.getActiveSlices());
+    return slices.stream().filter(this::isExpired).collect(Collectors.toSet());
+  }
+
+  private void deleteShard(final Slice s) {
+    deleteActor.delete(s);
+  }
+
+  /**
+   * An Inactive Shard is expired if it has not undergone a state change in the period of time
+   * defined by {@link InactiveShardRemover#ttlSeconds}. If it is expired, it is eligible for
+   * removal.
+   */
+  private boolean isExpired(final Slice slice) {
+
+    final String collectionName = slice.getCollection();
+    final String sliceName = slice.getName();
+
+    if (slice.getState() != Slice.State.INACTIVE) {
+      return false;
+    }
+
+    final String lastChangeTimestamp = slice.getStr(ZkStateReader.STATE_TIMESTAMP_PROP);
+    if (lastChangeTimestamp == null || lastChangeTimestamp.isEmpty()) {
+      log.warn(
+          "Collection {} Shard {} has no last change timestamp and will not be deleted",
+          collectionName,
+          sliceName);
+      return false;
+    }
+
+    final long epochTimestamp;
```

Comment:
I really wish time units were clearer here.  Is this Ns or Ms; just a variable name suffix would help.